# **EDA - Investigação de Churn**

Paulo Gustavo Duarte da Costa - IA & Dados 2026<br>
03 de Setembro de 2026

# **DataFrame**

**Telco Customer Churn - IBM**<br>
https://www.kaggle.com/datasets/blastchar/telco-customer-churn


### Investigue **Volume**, **Qualidade**, **Distribuição** e **Data readines**

#### Encontre **3 insights** e elabore **3 hipóteses** a partir da comparação entre **clientes que cancelaram e permaneceram**.


# **Entendimento de Negócio**

**Contexto**
> Uma empresa de telecomunicações percebeu que parte de seus clientes está **cancelando** seus serviços.

**Problema**
> Identificar clientes com maior **risco de cancelamento** e **comparar clientes que cancelaram e permaneceram**.

**Objetivo de Negócio**
>  A empresa quer verificar as **características de clientes que cancelaram**, buscando **estratégias para reduzir o percentual de cancelamento**.

**Objetivo de Ciência de Dados**
> Encontre **3 insights** e elabore **3 hipóteses** a partir da
comparação entre **clientes que cancelaram e permaneceram**.

**Stakeholdes**
> Gestores<br>
> Análistas<br>
> Equipe de Marketing

**Perguntas de Negócio**
> Qual é o perfil dos clientes que cancelam?<br>
> A taxa de cancelamento apresenta diferenças entre os gêneros?<br>
> Existe relação entre a quantidade de serviços contratados e a taxa de cancelamento?<br>
> Quais períodos apresentam maiores taxas de cancelamento?<br>
> Quais serviços estão mais associados ao cancelamento?<br>

**Métrica de Negócio**
- Identificar padrões nas características dos clientes que cancelaram e permaneceram
- Encontrar 3 insights relevantes a partir da comparação entre os grupos
- Elaborar 3 hipóteses que possam ser investigadas posteriormente


# **Entendimento dos Dados**

**DataFrame**
> Telco Customer Churn<br>
> https://www.kaggle.com/datasets/blastchar/telco-customer-churn

In [2]:
# Install dependencies as needed:
!pip install kagglehub[pandas-datasets]

In [3]:
import kagglehub
from kagglehub import KaggleDatasetAdapter
import pandas as pd
import plotly.express as px
from plotly.subplots import make_subplots
import matplotlib.pyplot as plt
import numpy as np
import plotly.graph_objects as go
import math
from sklearn.preprocessing import MinMaxScaler

In [4]:
# Set the path to the file you'd like to load
file_path = "WA_Fn-UseC_-Telco-Customer-Churn.csv"

# Load the latest version
df = kagglehub.load_dataset(
  KaggleDatasetAdapter.PANDAS,
  "blastchar/telco-customer-churn",
  file_path,
  # Provide any additional arguments like
  # sql_query or pandas_kwargs. See the
  # documenation for more information:
  # https://github.com/Kaggle/kagglehub/blob/main/README.md#kaggledatasetadapterpandas
)

/tmp/ipykernel_123473/560520164.py:5: DeprecationWarning: Use dataset_load() instead of load_dataset(). load_dataset() will be removed in a future version.
  df = kagglehub.load_dataset(


Using Colab cache for faster access to the 'telco-customer-churn' dataset.


In [5]:
# Dataframe
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [7]:
df.describe()

,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000


In [8]:
print("Linhas duplicadas: ", df.duplicated().sum())

Linhas duplicadas:  0


In [9]:
print("Colunas com valores faltantes\n")
print(df.isnull().sum())

Colunas com valores faltantes

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64


In [10]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [11]:
tabela = pd.DataFrame({
    "Coluna": df.columns,
    "Valores únicos": [df[col].nunique() for col in df.columns]
})

# Coluna Serviços
col_serv = tabela[tabela['Valores únicos'] <= 3]['Coluna'].tolist()


tabela.sort_values(by="Valores únicos", ascending=False)

,Coluna,Valores únicos
0,customerID,7043
19,TotalCharges,6531
18,MonthlyCharges,1585
5,tenure,73
17,PaymentMethod,4
14,StreamingMovies,3
12,TechSupport,3
10,OnlineBackup,3
13,StreamingTV,3
11,DeviceProtection,3


In [12]:
df.select_dtypes(include='object').replace(r'^\s*$', np.nan, regex=True).isnull().sum().sort_values(ascending=False).reset_index()

,index,0
0,TotalCharges,11
1,customerID,0
2,gender,0
3,Partner,0
4,PhoneService,0
5,Dependents,0
6,InternetService,0
7,OnlineSecurity,0
8,OnlineBackup,0
9,MultipleLines,0


In [13]:
df_can = df['Churn'].value_counts(ascending=False).reset_index()
df_can_per = df['Churn'].value_counts(normalize=True).reset_index()

fig1 = px.bar(
    df_can,
    x='Churn', y='count',
    labels={
        'Churn': 'Cancelou',
        'count': 'Quantidade'
    }
)


fig1.update_layout(
    title="Total de Cancelamentos",
    xaxis_title="Cancelou",
    yaxis_title="Quantidade",
    width=500,
    height=500
)

fig1.update_traces(
    texttemplate="%{y}",
    textposition="outside"
)


fig2 = px.pie(df_can_per, values='proportion', names='Churn', labels={'Churn': 'Cancelou', 'proportion': 'percentual'})

fig2.update_layout(
    title="Percentual de Cancelamentos",
    width=500,
    height=500
)

fig = make_subplots(
    rows=1,
    cols=2,
    subplot_titles=["Total de Cancelamentos", "Percentual de Cancelamentos"],
    specs=[[{'type':'xy'}, {'type':'domain'}]]
)

for trace in fig1.data:
    fig.add_trace(trace, row=1, col=1)

for trace in fig2.data:
    fig.add_trace(trace, row=1, col=2)


fig.update_layout(
    height=500,
    width=800,
    showlegend=True
)

fig.show()


**Entendendo o Dataframe**
> 7043 linhas com 21 colunas<br>
> Coluna TotalCharges precisa de tratamento<br>
> ~26.5% dos clientes fizeram cancelamento, entre eles:
> - ~13% de Homens<br>
> - ~13% de Mulheres<br>
>
> Quase 50% faz o cancelamento antes de 10 meses<br>
> Mais de 50% de cancelamento entre os 4 primeiros meses


# **Preparação dos Dados**


**Objetivos**
- Remover valores faltantes<br>
- Verificar e tratar outliers<br>
- Transformação de variavéis categoricas

> Transformando coluna TotalCharges em coluna numérica

In [14]:
df['TotalCharges'] = df['TotalCharges'].replace(' ', np.nan)
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')
df['TotalCharges'].head()

,TotalCharges
0,29.85
1,1889.50
2,108.15
3,1840.75
4,151.65


In [15]:
cols = df.select_dtypes(include='number').columns

cols = [
    col for col in cols
    if not df[col].dropna().isin([0, 1]).all()
]

n_colunas = 3
n_linhas = math.ceil(len(cols) / n_colunas)

fig = make_subplots(
    rows=n_linhas,
    cols=n_colunas,
    subplot_titles=cols
)

for i, col in enumerate(cols):
    linha = i // n_colunas + 1
    coluna = i % n_colunas + 1

    fig.add_trace(
        go.Box(
            y=df[col],
            name=col,
            showlegend=False
        ),
        row=linha,
        col=coluna
    )

fig.update_layout(
    height=400 * n_linhas,
    width=1000,
    showlegend=False
)

fig.show()


# **Insights e Hipóteses**

## **1. Insight**

**Váriavel**
> gender

**Pergunta**
> Gênero tem relação com cancelamento?

**Análise**

In [17]:
gen_churn = df.groupby(['Churn','gender'])['Churn'].count().reset_index(name='count')

fig = px.bar(gen_churn, x="gender", y="count", color='Churn', barmode='group', labels={'Churn': 'Cancelou', 'count': 'Quantidade', 'gender': 'Sexo'})

fig.update_layout(
    title="Total de cancelamentos por gênero",
    xaxis_title="Cancelou",
    yaxis_title="Quantidade",
    width=500,
    height=500
)

fig.update_traces(
    texttemplate="%{y}",
    textposition="outside"
)

fig.show()

In [18]:
gen_churn_to = df.groupby('gender')['Churn'].count().reset_index(name='count')

fig = px.bar(gen_churn_to, x="gender", y="count")

fig.update_layout(
    title="Total por gênero",
    xaxis_title="Gênero",
    yaxis_title="Quantidade",
    width=500,
    height=500
)

fig.update_traces(
    texttemplate="%{y}",
    textposition="outside"
)

fig.show()

In [19]:
# Porcentagem de cancelamento por gênero
homens = gen_churn_to[gen_churn_to['gender'] == 'Male']['count'].values[0]
mulheres = gen_churn_to[gen_churn_to['gender'] == 'Female']['count'].values[0]

homens_cancelados = gen_churn[gen_churn['gender'] == 'Male']['count'].values[1]
mulheres_canceladas = gen_churn[gen_churn['gender'] == 'Female']['count'].values[1]

homens_per = homens_cancelados / homens * 100
mulheres_per = mulheres_canceladas / mulheres * 100

print("Porcentagem de cancelamento por gênero:")
print(f"Homens: {homens_per:.2f}%")
print(f"Mulheres: {mulheres_per:.2f}%")
print(f"Diferenca: {(mulheres_per - homens_per):.2f}")

Porcentagem de cancelamento por gênero:
Homens: 26.16%
Mulheres: 26.92%
Diferenca: 0.76


**Resultado**<br>
Cancelamento por Gênero:
- De 3555 **Homens**, 930 cancelaram -> 26.16%
- De 3488 **Mulheres**, 939 cancelaram -> 26.92%
- Diferença -> **0.76%**

**Insight**
> **Gênero não apresentou uma diferença significativa para a chance de cancelamento.**

## **1. Hipótese**

**Hipótese**
> **Clientes com menores quantidades de meses com serviços contratados têm mais chance de cancelamento**

In [20]:
# Um histograma com a quantidade de clientes por meses cadastrados e que cancelaram.
fig = px.histogram(df, x="tenure", color="Churn", nbins=15, barmode="overlay", text_auto=True, opacity=0.75)

fig.update_layout(
    title="Quantidade de clientes que cancelaram ou não, por total de meses",
    yaxis_title="Contagem",
    width=1000,
    height=500,
)

fig.show()

In [21]:
# Um histograma com a quantidade de clientes por meses cadastrados e que cancelaram.
fig = px.histogram(df, x="tenure", color="Churn", nbins=8, barmode="overlay", text_auto=True, opacity=0.75)

fig.update_layout(
    title="Quantidade de clientes que cancelaram ou não, por total de meses",
    xaxis_title="Total de meses",
    yaxis_title="Contagem",
    width=1000,
    height=500,
)

fig.show()

### **Podemos observar pelos gráficos que clientes com menos de 10 meses tem maiores chances de cancelamento**

## **2. Insight**

**Váriavel**
> Quantiadade de serviços de clientes que cancelaram

**Pergunta**
> Quantidades de servicos com maiores números de clientes que cancelaram?

**Análise**

In [22]:
clientes_cancelaram = df[df['Churn'] == 'Yes'].copy()
colunas_texto = clientes_cancelaram.select_dtypes(include=['object', 'string']).columns

clientes_cancelaram[colunas_texto] = df[colunas_texto].replace({'Yes': 1, 'No': 0})
clientes_cancelaram.head(5)

/tmp/ipykernel_123473/2136663824.py:4: FutureWarning:

Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`



,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
2,3668-QPYBK,Male,0,0,0,2,1,0,DSL,1,...,0,0,0,0,Month-to-month,1,Mailed check,53.85,108.15,1
4,9237-HQITU,Female,0,0,0,2,1,0,Fiber optic,0,...,0,0,0,0,Month-to-month,1,Electronic check,70.70,151.65,1
5,9305-CDSKC,Female,0,0,0,8,1,1,Fiber optic,0,...,1,0,1,1,Month-to-month,1,Electronic check,99.65,820.50,1
8,7892-POOKP,Female,0,1,0,28,1,1,Fiber optic,0,...,1,1,1,1,Month-to-month,1,Electronic check,104.80,3046.05,1
13,0280-XJGEX,Male,0,0,0,49,1,1,Fiber optic,0,...,1,0,1,1,Month-to-month,1,Bank transfer (automatic),103.70,5036.30,1


In [23]:
colunas_servicos_reais = [
    'PhoneService', 'MultipleLines', 'OnlineSecurity',
    'OnlineBackup', 'DeviceProtection', 'TechSupport',
    'StreamingTV', 'StreamingMovies'
]

servicos_numericos = clientes_cancelaram[colunas_servicos_reais].apply(pd.to_numeric, errors='coerce').fillna(0)
clientes_cancelaram['total_servicos'] = servicos_numericos.sum(axis=1)

# Agrupando os dados para contar a quantidade de clientes por número de serviços contratados
cli = clientes_cancelaram[clientes_cancelaram['total_servicos'] > 0].copy()
servicos_agrupados = cli['total_servicos'].value_counts().reset_index(name='count')
servicos_agrupados.columns = ['total_servicos', 'count']
servicos_agrupados = servicos_agrupados.sort_values(by='total_servicos')

# Criando o gráfico de barras
fig = px.bar(
    servicos_agrupados,
    x="total_servicos",
    y="count",
    text_auto=True,
    opacity=0.75
)

fig.update_layout(
    title="Total de serviços de clientes que cancelaram",
    xaxis_title="Quantidade de Serviços Contratados",
    yaxis_title="Quantidade de Clientes (Contagem)",
    width=1000,
    height=500
)

fig.update_traces(
    textposition="outside"
)

fig.show()

**Resultado**<br>
De 1 - 3 serviços têm a maior quantidade dos clientes que cancelaram.

**Insight**
> **Clientes que cancelaram têm a tendência de terem menos serviços solicitados.**

## **2.Hipótese**

**Hipótese**
> **Clientes com valores mensais maiores têm maior taxa de permanência**

In [24]:
df_copy = df[['customerID', 'MonthlyCharges', 'Churn']].copy()
df_copy.head()

,customerID,MonthlyCharges,Churn
0,7590-VHVEG,29.85,No
1,5575-GNVDE,56.95,No
2,3668-QPYBK,53.85,Yes
3,7795-CFOCW,42.30,No
4,9237-HQITU,70.70,Yes


In [25]:
fig = px.histogram(df_copy, x="MonthlyCharges", color="Churn", nbins=8, barmode="overlay", text_auto=True, opacity=0.75)

fig.update_layout(
    title="Valor mensal de clientes",
    xaxis_title="Valor",
    yaxis_title="Contagem",
    width=1000,
    height=500,
)

fig.show()

### **Podemos observar pelo gráfico que essa hipótese não é valida, pois a maior taxa de permanência está com clientes que têm custo mensal menor que R$60.**

## **3. Insight**

**Váriavel**
> Valor mensal com valor total pago

**Pergunta**
> Tem relação entre valor mensal e valor total pago de clientes que cancelaram e dos que permaneceram?

**Análise**

In [26]:
df_analis = df[['MonthlyCharges', 'TotalCharges','Churn']].copy()
colunas = ['MonthlyCharges', 'TotalCharges']

scaler = MinMaxScaler()

df_analis[colunas] = scaler.fit_transform(df_analis[colunas])
df_analis.head()

,MonthlyCharges,TotalCharges,Churn
0,0.115423,0.001275,No
1,0.385075,0.215867,No
2,0.354229,0.010310,Yes
3,0.239303,0.210241,No
4,0.521891,0.015330,Yes


In [27]:
fig = px.scatter(
    df_analis,
    x="MonthlyCharges",
    y="TotalCharges",
    color="Churn",
    color_discrete_map={
        "Yes": "red",
        "No": "green"
    },
    opacity=0.4,
    labels={
        "MonthlyCharges": "Valor Mensal",
        "TotalCharges": "Total Pago",
        "Churn": "Cancelou"
    },
    title="Relação entre Valor Mensal e Total Pago"
)

fig.update_layout(
    width=800,
    height=800
)


fig.show()

In [28]:
df_yes = df_analis[df_analis['Churn'] == 'Yes'].copy()

print("Correlação de valor mensal com total pago de clientes que fizeram cancelamento\n")
print(df_yes[colunas].corr())

Correlação de valor mensal com total pago de clientes que fizeram cancelamento

                MonthlyCharges  TotalCharges
MonthlyCharges        1.000000      0.545772
TotalCharges          0.545772      1.000000


In [29]:
df_yes = df_analis[df_analis['Churn'] == 'No'].copy()

print("Correlação de valor mensal com total pago que não fizeram cancelamento\n")
print(df_yes[colunas].corr())

Correlação de valor mensal com total pago que não fizeram cancelamento

                MonthlyCharges  TotalCharges
MonthlyCharges         1.00000       0.75711
TotalCharges           0.75711       1.00000


**Resultado**<br>
Pelo gráfico podemos observar que os valores parecem estar bem dispersos, mas quando fazemos o cálculo da correlação:
- Para clientes que **não fizeram cancelamento** -> **Correlação de 0.76** aproximadamente, com uma **correlação positiva forte**.
- Para cliente que **fizeram cancelamento** -> **Correlação de 0.54** aproximadamente, com uma **correlação positiva moderada**.

**Insight**
> **Esse comportamento sugere que a relação entre o valor mensal e o total pago varia de acordo com o grupo de cancelamento, mas que essa relação é mais forte em relação aos que não cancelaram.**

## **3.Hipótese**

**Hipótese**
> **A taxa de cancelamento varia de acordo com o método de pagamento utilizado pelos clientes.**

In [30]:
df_can = df[['PaymentMethod', 'Churn']].copy()
df_can.head()

,PaymentMethod,Churn
0,Electronic check,No
1,Mailed check,No
2,Mailed check,Yes
3,Bank transfer (automatic),No
4,Electronic check,Yes


In [31]:
contagem = df_can.groupby(['PaymentMethod','Churn'])['Churn'].count().rename('Contagem')
contagem = contagem.reset_index()

contagem.head()

,PaymentMethod,Churn,Contagem
0,Bank transfer (automatic),No,1286
1,Bank transfer (automatic),Yes,258
2,Credit card (automatic),No,1290
3,Credit card (automatic),Yes,232
4,Electronic check,No,1294


In [32]:
fig = px.bar(
    contagem,
    x='PaymentMethod',
    y='Contagem',
    color='Churn',
    barmode='group',
    text='Contagem',
    labels={
        'PaymentMethod': 'Método de pagamento',
        'Contagem': 'Quantidade de clientes',
        'Churn': 'Cancelamento'
    },
    title='Quantidade de clientes por método de pagamento e cancelamento'
)

fig.show()

In [33]:
check_ele = 1071 / (2365 / 100)
check_ele

45.28541226215645

### **Todos métodos de cancelamento tem taxas parecidas, menos Check Eletônico, que apresenta uma grande quantidade de cancelamento, de aproximadamente 45% do total de cliente com método de pagamento como Check eletrônico.**